### PR TITLE
Chapter 7 Readme and Example (English version)

### PR DESCRIPTION
**Pull request by Ariadne330**
During the verification of Fine-tune Process, I ran into some issues:
- the `kernel die` problem in Jupyter is just as the description in [this issue](https://github.com/analytics-zoo/nano/issues/553), so I wrote the example in Markdown style
- the `export_merged_model.py` could not run successfully firstly, and with ruonan's help I fix this by changing the version of `accelerate`, recorded in [this issue](https://github.com/intel-analytics/BigDL/issues/9312)

And Chinese version should be added after fixing problems in English example

### CODE REVIEW

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+```python
+import intel_extension_for_pytorch as ipex
+```
+
+## 7.1.2 QLoRA Finetuning
+
+### 7.1.2.1 Load Model in Low Precision
+
+
+A popular open-source LLM [meta-llama/Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-hf) is chosen to illustrate the process of model finetuning.
```
Comment (Oscilloscope98) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:31]: of QLoRA Finetuning?

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+To help you better understand the finetuning process, in this tutorial, we provide a practical guide leveraging BigDL-LLM to tune a large language model to a specific task.  [Llama-2-7b-hf](https://huggingface.co/meta-llama/Llama-2-7b-hf) is used as an example here to adapt the text generation implementation.
+
+## 7.1.1 Enable BigDL-LLM on Intel GPUs
+
+
+### 7.1.1.1 Install BigDL-LLM on Intel GPUs
+
+After following the steps in [Readme](./README.md#70-environment-setup)  to set up the environment, you can install BigDL-LLM in terminal with the command below:
+```bash
+pip install bigdl-llm[xpu] -f https://developer.intel.com/ipex-whl-stable-xpu
```
Comment (Oscilloscope98) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:12]: Some dependencies are missed here. Please refer to [here](https://github.com/intel-analytics/BigDL/tree/main/python/llm/example/GPU/QLoRA-FineTuning#1-install)

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+```
+> **Note**
+>
+> Instead of `from peft import prepare_model_for_kbit_training, get_peft_model` as we did for regular QLoRA using bitandbytes and cuda, we import them from `bigdl.llm.transformers.qlora` here to get a BigDL-LLM compatible PEFT model. And the rest is just the same as regular LoRA finetuning process using `peft`.
+>
+
+> **Note**
+>
+> For Llama 2 (7B) model, the `target_modules` is set as `["q_proj", "k_proj", "v_proj"]`.
+> You can find a list of **target_modules available** on the [Hugging Face Documentation](https://github.com/huggingface/peft/blob/39ef2546d5d9b8f5f8a7016ec10657887a867041/src/peft/utils/other.py#L220).
```
Comment (Oscilloscope98) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:88]: Is there any reason to specify to this specific commit?

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+data = data.map(lambda samples: tokenizer(samples["quote"]), batched=True)
+```
+
+> **Note**
+>
+> The dataset path here is default to be Huggingface repo id. 
+> If you have already downloaded the `.jsonl` file from [Abirate/english_quotes](https://huggingface.co/datasets/Abirate/english_quotes/blob/main/quotes.jsonl), you could use `data = load_dataset("json", data_files= "path/to/your/.jsonl/file")` to specify the local path instead of `data = load_dataset("Abirate/english_quotes")`.
+
+### 7.1.2.4 Load Tokenizer
+A tokenizer is also needed for LLM inference. You can use [Huggingface transformers](https://huggingface.co/docs/transformers/index) API to load the tokenizer directly. It can be used seamlessly with models loaded by BigDL-LLM. For Llama 2, the corresponding tokenizer class is `LlamaTokenizer`.
```
Comment (Oscilloscope98) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:105]: I feel like it is not clear here why we need tokenizer for QLoRA finetuning :)

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        adapter_path,
+        device_map={"": "cpu"},
+        torch_dtype=torch.float16,
+    )
+
+
+```
+> **Note**
+>
+> Instead of from peft import PeftModel, we import PeftModel from bigdl.llm.transformers.qlora as a BigDL-LLM compatible model.
```
Comment (Oscilloscope98) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:208]:
```
Instead of `from peft import PeftModel`, we `import PeftModel from bigdl.llm.transformers.qlora` as a BigDL-LLM compatible model.
```

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+base_model.save_pretrained(output_path, state_dict=deloreanized_sd)
+tokenizer.save_pretrained(output_path)
+
+```
+
+
+## 7.1.4 Inference with Fine-tuned model
+After merging and deploying the models, we can test the performance of the fine-tuned model. 
+The detailed instructions of running
+LLM inference with BigDL-LLM optimizations could be found in [Chapter 6](../ch_6_GPU_Acceleration/6_1_GPU_Llama2-7B.md), here we quickly go through the preparation of model inference.
```
Comment (Oscilloscope98) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:252]: why changing to a new line here?

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+
+### 7.1.4.1 Inference with the Fine-tuned Model
+```python
+model_path = "./outputs/checkpoint-200-merged"
+model = AutoModelForCausalLM.from_pretrained(pretrained_model_name_or_path = model_path,load_in_4bit=True)
+model.config.use_cache = True
```
Comment (Oscilloscope98) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:258]: Do we need explicitly emphasize here `use_cache=True`?

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+    output = model.generate(input_ids, max_new_tokens=32)
+    output_str = tokenizer.decode(output[0], skip_special_tokens=True)
+    print(output_str)
+```
+Here are the inference results: 
+```
+The paradox of time and eternity is that, on the one hand, we experience time as linear and progressive, and on the other hand, we experience time as cyclical. And the
+```
+
+### 7.1.4.2 Inference with the Pre-trained Model
```
Comment (Oscilloscope98) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:279]: We could directly show the difference between the results of fine-tuned model and the pretrained model. No need to show the inference code again for the pretrained model. It may confuse the reader :)

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+model = prepare_model_for_kbit_training(model)
+```
+
+Next, we can obtain a PEFT model from the optimized model and a configuration object containing the parameters as follows:  
+
+```python
+from bigdl.llm.transformers.qlora import get_peft_model
+from peft import LoraConfig
+
+config = LoraConfig(r=8, # defines how many parameters will be trained
```
Comment (plusbang) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:71]: Please  update comments of these parameters, referring to https://huggingface.co/docs/peft/conceptual_guides/lora#common-lora-parameters-in-peft

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+{'loss': 1.2266, 'learning_rate': 0.00015555555555555556, 'epoch': 0.1}                                              
+{'loss': 1.1534, 'learning_rate': 0.00013333333333333334, 'epoch': 0.13}                                             
+{'loss': 0.9368, 'learning_rate': 0.00011111111111111112, 'epoch': 0.16}                                             
+{'loss': 0.9321, 'learning_rate': 8.888888888888889e-05, 'epoch': 0.19}                                              
+{'loss': 0.9902, 'learning_rate': 6.666666666666667e-05, 'epoch': 0.22}                                              
+{'loss': 0.8593, 'learning_rate': 4.4444444444444447e-05, 'epoch': 0.26}                                             
+{'loss': 1.0055, 'learning_rate': 2.2222222222222223e-05, 'epoch': 0.29}                                             
+{'loss': 1.0081, 'learning_rate': 0.0, 'epoch': 0.32}                                                                
+{'train_runtime': 285.5368, 'train_samples_per_second': 2.802, 'train_steps_per_second': 0.7, 'train_loss': 1.1155566596984863, 'epoch': 0.32}
```
Comment (plusbang) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:158]: Should we hide the specific value of `train_runtime`, `train_samples_per_second` and `train_steps_per_second`

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+    )
+
+
+```
+> **Note**
+>
+> Instead of from peft import PeftModel, we import PeftModel from bigdl.llm.transformers.qlora as a BigDL-LLM compatible model. 
+> 
+> **Note**
+> The adapter path is the local path you save the fine-tuned model, in our case is `./outputs/checkpoint-200`.
```
Comment (plusbang) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:211]: Please check it as above mentioned.

File: ch_7_Finetune/README.md [hunk 12]
```diff
@@ -1 +1,60 @@
...
+
+### 7.0.4 Best Known Configuration on Linux
+
+For optimal performance on Intel GPUs, it is recommended to set several environment variables:
+
+```bash
+# configure OneAPI environment variables
+source /opt/intel/oneapi/setvars.sh
+
+export USE_XETLA=OFF
```
[the remark this answers is not in the document]
Comment (Oscilloscope98) [reply ch_7_Finetune/README.md:55]: I guess it is because we also do inference on the fine-tuned model in this tutorial, so the two env param are set. Or should we add some more comments for explanation/separate the env setting here.

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md (hunk 7 above)
Comment (Ariadne330) [reply ch_7_Finetune/7_1_Finetune_Llama2-7B.md:258]: According to [the example](https://github.com/intel-analytics/BigDL/blob/main/python/llm/example/GPU/QLoRA-FineTuning/qlora_finetuning.py), this line can be removed here

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md (hunk 6 above)
Comment (Ariadne330) [reply ch_7_Finetune/7_1_Finetune_Llama2-7B.md:252]: ( it was an accident・ˍ・

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 13]
```diff
@@ -0,0 +1,321 @@
...
+{'loss': 0.9902, 'learning_rate': 6.666666666666667e-05, 'epoch': 0.22}                                              
+{'loss': 0.8593, 'learning_rate': 4.4444444444444447e-05, 'epoch': 0.26}                                             
+{'loss': 1.0055, 'learning_rate': 2.2222222222222223e-05, 'epoch': 0.29}                                             
+{'loss': 1.0081, 'learning_rate': 0.0, 'epoch': 0.32}                                                                
+{'train_runtime': 285.5368, 'train_samples_per_second': 2.802, 'train_steps_per_second': 0.7, 'train_loss': 1.1155566596984863, 'epoch': 0.32}
+100%|██████████████████████████████████████████████████████████████████████████████| 200/200 [04:45<00:00,  1.43s/it]
+TrainOutput(global_step=200, training_loss=1.1155566596984863, metrics={'train_runtime': 285.5368, 'train_samples_per_second': 2.802, 'train_steps_per_second': 0.7, 'train_loss': 1.1155566596984863, 'epoch': 0.32})
+```
+The final LoRA weights and configurations have been saved to `${output_dir}/checkpoint-{max_steps}`, which can be used for merging.
```
[the remark this answers is not in the document]
Comment (Ariadne330) [reply ch_7_Finetune/7_1_Finetune_Llama2-7B.md:162]: fix it to `${output_dir}/checkpoint-{max_steps}/adapter_model.bin` and `${output_dir}/checkpoint-{max_steps}/adapter_config.json`

File: ch_7_Finetune/README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (Ariadne330) [reply ch_7_Finetune/README.md:55]: Is it ok to state this part as optional? ( Currently I change it in this way

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md (hunk 3 above)
Comment (Ariadne330) [reply ch_7_Finetune/7_1_Finetune_Llama2-7B.md:88]: Just follow another comprehensive tutorial to write this part ( Should I remove it?

File: ch_7_Finetune/7_1_Finetune_Llama2-7B.md [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+pip install peft==0.5.0
+pip install accelerate==0.23.0
+```
+
+> **Note**
+> The above command will install `intel_extension_for_pytorch==2.0.110+xpu` as default
+
+### 7.1.1.2 Import `intel_extension_for_pytorch`
+
+After installation, let's move to the **Python scripts** of this tutorial. First of all you need to import `intel_extension_for_pytorch` first for BigDL-LLM optimizations:
```
Comment (yangw1234) [inline ch_7_Finetune/7_1_Finetune_Llama2-7B.md:23]: import `intel_extension_for_pytorch` for using xpu device?

File: ch_7_Finetune/README.md [hunk 15]
```diff
@@ -1 +1,68 @@
...
+
+```bash
+# Python 3.9 is recommended for running BigDL-LLM
+conda create -n llm-finetune python=3.9 
+conda activate llm-finetune 
+```
+
+### 7.0.4 Set OneAPI Environment Variables
+
+The following line can help you set OneAPI environment variables.
```
Comment (Oscilloscope98) [inline ch_7_Finetune/README.md:51]: You need to set OneAPI environment variables for BigDL-LLM on Intel GPUs.

**Review by Oscilloscope98 — APPROVED** [review]
Merge for now for the English version. May raise other PRs for any other comments :)
